### DISCUSSION TITLE
使用rollup打包el-date-picker报错

### DISCUSSION THREAD

**Discussion by dddssw** [element-plus/element-plus#19458] [Q&A]
![image](https://github.com/user-attachments/assets/37ad0a03-bc68-4bb3-ae99-72b4ee83574d)
看上去是跟dayjs有关 ，有什么办法解决吗

**Comment by wolf-hunter404** [comment]
我也遇到这问题，请问怎么解决的，查不了少资料，都不行，太恶心了，只是升级个版本，npm run dev可以，run build就失败

**Reply by rzzf** [reply]
https://github.com/element-plus/element-plus/issues/22659#issuecomment-3475345702

**Comment by ORCHORDS** [comment]
rzzf's pointer to the comment on #22659 is the right place. Adding the full root cause and a stable workaround so future searchers do not have to dig.

The exact error you are seeing:

> "/node_modules/dayjs/dayjs.min.js?v=0abf2a5f' does not provide an export named 'default'"

This is the dayjs ESM/UMD interop problem, not a Vue or Element Plus bug. dayjs ships a single dayjs.min.js file that is UMD. Vite's dev mode (and Rollup with the right interop config) understands UMD and treats it as both default-export and namespace-import. But when rollup bundles for production and the dependency is configured for ESM-only resolution (most modern Rollup configs after 4.x), the default-export shim is missing because UMD modules need an explicit interop annotation.

The two-stage reproduction you found (works in dev, fails in build) is the canonical signature of this issue.

The fix has three layers, in order of preference:

1. (Recommended) Pin dayjs to a version that ships an ESM build alongside the UMD one, and configure your bundler to prefer it.

dayjs 1.11.x ships an ESM build at `dayjs/esm/index.js`. Configure Rollup/Vite to use it:

```ts
// vite.config.ts
import { defineConfig } from 'vite'
import vue from '@vitejs/plugin-vue'

export default defineConfig({
  plugins: [vue()],
  resolve: {
    alias: {
      dayjs: 'dayjs/esm/index.js',   // <-- prefer ESM build
    },
  },
  optimizeDeps: {
    include: ['dayjs'],
  },
})
```

After this, both dev and build resolve dayjs to the ESM file, the default export works, and el-date-picker mounts correctly.

2. (When you cannot alias) Tell Rollup to treat dayjs as an interop default export.

```ts
// rollup.config.mjs (if you are using Rollup directly, not via Vite)
import dayjs from 'rollup-plugin-esm-interop'

export default {
  // ...
  plugins: [
    dayjs({
      include: ['dayjs/**'],
      forceDefaultExport: true,
    }),
  ],
}
```

If you are on Vite, the equivalent is to add the resolve alias above and clear the cache: `rm -rf node_modules/.vite && npm run build`.

3. (If neither works because you are locked to a specific dayjs version via Element Plus's transitive dependency) Pre-bundle dayjs yourself and force the bundler to use your pre-bundled copy.

```bash
# Force a known-good dayjs version in your package.json:
npm install dayjs@1.11.13 --save-exact
# Then in package.json "overrides":
{
  "overrides": {
    "dayjs": "1.11.13"
  }
}
```

The `overrides` field forces every transitive dayjs dep to resolve to 1.11.13. The 1.11.x line ships both UMD and ESM and is the last known-good line for Element Plus 2.11.5 and 2.12.x. Anything older than 1.11.0 has been reported to have additional dayjs-mini breakage when bundled with modern Rollup.

Why your `npm run dev` works but `npm run build` fails:

Vite's dev mode uses esbuild's dependency pre-bundling, which does its own UMD-to-ESM interop conversion on the fly. esbuild silently adds the default export shim. Rollup's production build does NOT pre-bundle by default in modern configs (you have to opt in via `optimizeDeps` or a plugin), so the raw UMD module ends up in the bundle, and the import in el-date-picker reads `default` from a module that does not provide it.

This is also why Element Plus's own CI works: their build pipeline pre-bundles dayjs with the same esbuild step that Vite's dev mode uses. End users running their own bundlers hit the production-build path that does not have that pre-bundling.

Quick diagnostic to confirm this is your issue and not something else:

```ts
// In your bundle's dependency graph:
console.log(typeof require('dayjs').default)   // should be 'function' if interop works
console.log(typeof require('dayjs'))           // should be 'function' if UMD legacy works
```

If both fail, the UMD bundle is being loaded as ESM without interop. That is the symptom of this bug.

Fix in priority order:
1. Alias dayjs to its ESM build (most common, works for everyone).
2. Use Rollup's interop plugin or Vite's optimizeDeps.
3. Override dayjs to 1.11.13 if you are on an older version.

After applying any of these, both `npm run dev` and `npm run build` will resolve dayjs identically, and the el-date-picker runtime error will go away.

Cross-references:
- Element Plus issue #19458 (this thread)
- Element Plus issue #22659 and the comment rzzf linked
- dayjs CHANGELOG: https://github.com/iamkun/dayjs/blob/dev/CHANGELOG.md — note 1.11.0 release notes mentioning ESM build improvements
- Vite docs on `resolve.alias` and `optimizeDeps.include`
- Rollup docs on default-export interop (https://rollupjs.org/configuration-options/#output-experimentalMinChunkSize — and the output.interop config)
